src/EFCore.Relational/Query/Internal/RelationalValueConverterCompensatingExpressionVisitor.cs

```diff
@@ -142,9 +142,9 @@ private Expression VisitLeftJoin(LeftJoinExpression leftJoinExpression)
     [return: NotNullIfNotNull("sqlExpression")]
     private SqlExpression? TryCompensateForBoolWithValueConverter(SqlExpression? sqlExpression)
     {
-        if (sqlExpression is ColumnExpression columnExpression
-            && columnExpression.TypeMapping!.ClrType == typeof(bool)
-            && columnExpression.TypeMapping.Converter != null)
+        if ((sqlExpression is ColumnExpression || sqlExpression is JsonScalarExpression)
+            && sqlExpression.TypeMapping!.ClrType == typeof(bool)
+            && sqlExpression.TypeMapping.Converter != null)
```

Comment:
In case you haven't noticed I'm really excited about pattern matching lately ;)

```suggestion
        if (sqlExpression is ColumnExpression or JsonScalarExpression { TypeMapping.Converter: not null }
            && sqlExpression.TypeMapping!.ClrType == typeof(bool))
```

Reply:
not worth in this case tho, `sqlExpression.TypeMapping.Converter != null` needs to apply to both types

Reply:
It's just a nit, up to you :) Though doing `sqlExpression is ColumnExpression or JsonScalarExpression` does seem simply better than `sqlExpression is ColumnExpression || sqlExpression is JsonScalarExpression`.

Reply:
oh yeah, that one i can get behind :)